File: examples/go/producer.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/segmentio/kafka-go"
)

// TelemetryEvent represents an LLM telemetry event
type TelemetryEvent struct {
	Timestamp        string                 `json:"timestamp"`
	ServiceName      string                 `json:"service_name"`
	ModelName        string                 `json:"model_name"`
	LatencyMs        float64                `json:"latency_ms"`
	PromptTokens     int                    `json:"prompt_tokens"`
	CompletionTokens int                    `json:"completion_tokens"`
	TotalTokens      int                    `json:"total_tokens"`
	CostUsd          float64                `json:"cost_usd"`
	UserID           string                 `json:"user_id"`
	SessionID        string                 `json:"session_id"`
	RequestID        string                 `json:"request_id"`
	PromptText       string                 `json:"prompt_text,omitempty"`
	ResponseText     string                 `json:"response_text,omitempty"`
	Metadata         map[string]interface{} `json:"metadata,omitempty"`
}

// TelemetryProducer sends LLM telemetry events to Kafka
type TelemetryProducer struct {
	writer *kafka.Writer
	topic  string
}

// NewTelemetryProducer creates a new telemetry producer
func NewTelemetryProducer(brokers []string, topic string) *TelemetryProducer {
	writer := &kafka.Writer{
		Addr:         kafka.TCP(brokers...),
		Topic:        topic,
		Balancer:     &kafka.LeastBytes{},
		RequiredAcks: kafka.RequireAll,
		MaxAttempts:  3,
		WriteTimeout: 10 * time.Second,
		ReadTimeout:  10 * time.Second,
	}

	log.Printf("Connected to Kafka brokers: %v", brokers)
	return &TelemetryProducer{
		writer: writer,
		topic:  topic,
	}
}

// CreateTelemetryEvent creates a telemetry event
func (p *TelemetryProducer) CreateTelemetryEvent(
	serviceName, modelName string,
	latencyMs float64,
	promptTokens, completionTokens int,
	costUsd float64,
	userID, sessionID string,
	metadata map[string]interface{},
) TelemetryEvent {
	requestID := fmt.Sprintf("req-%d-%d", time.Now().UnixMilli(), rand.Intn(10000))

	return TelemetryEvent{
		Timestamp:        time.Now().UTC().Format(time.RFC3339Nano),
		ServiceName:      serviceName,
		ModelName:        modelName,
		LatencyMs:        latencyMs,
		PromptTokens:     promptTokens,
		CompletionTokens: completionTokens,
		TotalTokens:      promptTokens + completionTokens,
		CostUsd:          costUsd,
		UserID:           userID,
		SessionID:        sessionID,
		RequestID:        requestID,
		Metadata:         metadata,
	}
}

// SendEvent sends a telemetry event to Kafka
func (p *TelemetryProducer) SendEvent(ctx context.Context, event TelemetryEvent) error {
	value, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	msg := kafka.Message{
		Key:   []byte(event.RequestID),
		Value: value,
		Time:  time.Now(),
	}

	err = p.writer.WriteMessages(ctx, msg)
	if err != nil {
		return fmt.Errorf("failed to send event: %w", err)
	}

	log.Printf("Sent event %s to topic %s", event.RequestID, p.topic)
	return nil
}

// Close closes the producer
func (p *TelemetryProducer) Close() error {
	return p.writer.Close()
}

// SimulateNormalTraffic generates normal LLM traffic patterns
func SimulateNormalTraffic(ctx context.Context, producer *TelemetryProducer, numEvents int) {
	log.Printf("Simulating %d normal traffic events...", numEvents)

	models := []string{"gpt-4", "gpt-3.5-turbo", "claude-3-opus", "claude-3-sonnet"}
	services := []string{"chat-api", "completion-api", "assistant-api"}
	regions := []string{"us-east-1", "us-west-2", "eu-west-1"}

	for i := 0; i < numEvents; i++ {
		select {
		case <-ctx.Done():
			return
		default:
		}

		// Normal latency: 500-3000ms
		latencyMs := 500.0 + rand.Float64()*2500.0

		// Normal token counts
		promptTokens := 50 + rand.Intn(450)
		completionTokens := 100 + rand.Intn(700)

		// Calculate cost (example pricing)
		model := models[rand.Intn(len(models))]
		var costUsd float64
		if strings.Contains(model, "gpt-4") {
			costUsd = float64(promptTokens)*0.00003 + float64(completionTokens)*0.00006
		} else {
			costUsd = float64(promptTokens)*0.000001 + float64(completionTokens)*0.000002
		}

		event := producer.CreateTelemetryEvent(
			services[rand.Intn(len(services))],
			model,
			latencyMs,
			promptTokens,
			completionTokens,
			costUsd,
			fmt.Sprintf("user-%d", rand.Intn(100)),
			fmt.Sprintf("session-%d", rand.Intn(50)),
			map[string]interface{}{
				"region":      regions[rand.Intn(len(regions))],
				"api_version": "v1",
			},
		)

		if err := producer.SendEvent(ctx, event); err != nil {
			log.Printf("Error sending event: %v", err)
		}

		time.Sleep(100 * time.Millisecond)
	}
}

// SimulateAnomalousTraffic generates anomalous LLM traffic patterns
func SimulateAnomalousTraffic(ctx context.Context, producer *TelemetryProducer, numEvents int) {
	log.Printf("Simulating %d anomalous traffic events...", numEvents)

	anomalyTypes := []struct {
		Type        string
		Description string
	}{
		{"high_latency", "Extremely high latency"},
		{"high_tokens", "Unusually high token count"},
		{"high_cost", "Abnormally high cost"},
		{"suspicious_pattern", "Suspicious usage pattern"},
	}

	for i := 0; i < numEvents; i++ {
		select {
		case <-ctx.Done():
			return
		default:
		}

		anomaly := anomalyTypes[rand.Intn(len(anomalyTypes))]
		var latencyMs float64
		var promptTokens, completionTokens int

		switch anomaly.Type {
		case "high_latency":
			// Anomalous: 20-60 seconds
			latencyMs = 20000.0 + rand.Float64()*40000.0
			promptTokens = 100 + rand.Intn(400)
			completionTokens = 200 + rand.Intn(600)

		case "high_tokens":
			// Anomalous: very high token count
			latencyMs = 5000.0 + rand.Float64()*10000.0
			promptTokens = 5000 + rand.Intn(10000)
			completionTokens = 8000 + rand.Intn(12000)

		case "high_cost":
			// Anomalous: extremely high cost
			latencyMs = 8000.0 + rand.Float64()*12000.0
			promptTokens = 8000 + rand.Intn(7000)
			completionTokens = 10000 + rand.Intn(15000)

		default: // suspicious_pattern
			// Multiple rapid requests from same user
			latencyMs = 1000.0 + rand.Float64()*2000.0
			promptTokens = 50 + rand.Intn(150)
			completionTokens = 50 + rand.Intn(150)
		}

		costUsd := float64(promptTokens)*0.00003 + float64(completionTokens)*0.00006

		event := producer.CreateTelemetryEvent(
			"chat-api",
			"gpt-4",
			latencyMs,
			promptTokens,
			completionTokens,
			costUsd,
			"user-suspicious",
			fmt.Sprintf("session-anomaly-%d", i),
			map[string]interface{}{
				"anomaly_type": anomaly.Type,
				"description":  anomaly.Description,
				"simulated":    true,
			},
		)

		if err := producer.SendEvent(ctx, event); err != nil {
			log.Printf("Error sending event: %v", err)
		}

		log.Printf("Sent anomalous event: %s", anomaly.Type)
		time.Sleep(500 * time.Millisecond)
	}
}

func main() {
	brokersFlag := flag.String("brokers", "localhost:9092", "Comma-separated list of Kafka brokers")
	topicFlag := flag.String("topic", "llm.telemetry", "Kafka topic name")
	normalEvents := flag.Int("normal-events", 20, "Number of normal events to generate")
	anomalousEvents := flag.Int("anomalous-events", 5, "Number of anomalous events to generate")
	continuous := flag.Bool("continuous", false, "Run continuously")
	flag.Parse()

	rand.Seed(time.Now().UnixNano())

	brokers := strings.Split(*brokersFlag, ",")
	producer := NewTelemetryProducer(brokers, *topicFlag)
	defer producer.Close()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Handle graceful shutdown
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-sigChan
		log.Println("Received interrupt signal, shutting down...")
		cancel()
	}()

	if *continuous {
		log.Println("Running in continuous mode (Ctrl+C to stop)...")
		for {
			select {
			case <-ctx.Done():
				log.Println("Shutting down...")
				return
			default:
				SimulateNormalTraffic(ctx, producer, *normalEvents)
				SimulateAnomalousTraffic(ctx, producer, *anomalousEvents)
				log.Println("Waiting 10 seconds before next batch...")
				time.Sleep(10 * time.Second)
			}
		}
	} else {
		SimulateNormalTraffic(ctx, producer, *normalEvents)
		SimulateAnomalousTraffic(ctx, producer, *anomalousEvents)
		log.Println("Finished generating events")
	}
}

File: examples/python/producer.py
#!/usr/bin/env python3
"""
LLM-Sentinel Python Telemetry Producer Example

This example demonstrates how to send LLM telemetry events to Kafka
for processing by LLM-Sentinel.

Requirements:
    pip install kafka-python

Usage:
    python producer.py --brokers localhost:9092 --topic llm.telemetry
"""

import argparse
import json
import logging
import random
import time
from datetime import datetime, timezone
from typing import Dict, Any
from kafka import KafkaProducer

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class LLMTelemetryProducer:
    """Producer for sending LLM telemetry events to Kafka."""

    def __init__(self, brokers: list[str], topic: str):
        """Initialize the telemetry producer.

        Args:
            brokers: List of Kafka broker addresses
            topic: Kafka topic name
        """
        self.topic = topic
        self.producer = KafkaProducer(
            bootstrap_servers=brokers,
            value_serializer=lambda v: json.dumps(v).encode('utf-8'),
            acks='all',
            retries=3,
            max_in_flight_requests_per_connection=1,
        )
        logger.info(f"Connected to Kafka brokers: {brokers}")

    def create_telemetry_event(
        self,
        service_name: str,
        model_name: str,
        latency_ms: float,
        prompt_tokens: int,
        completion_tokens: int,
        cost_usd: float,
        user_id: str = "user-123",
        session_id: str = "session-456",
        request_id: str = None,
        prompt_text: str = None,
        response_text: str = None,
        metadata: Dict[str, Any] = None,
    ) -> Dict[str, Any]:
        """Create a telemetry event payload.

        Args:
            service_name: Name of the service generating the event
            model_name: LLM model name (e.g., "gpt-4", "claude-3")
            latency_ms: Request latency in milliseconds
            prompt_tokens: Number of tokens in the prompt
            completion_tokens: Number of tokens in the completion
            cost_usd: Cost of the request in USD
            user_id: User identifier
            session_id: Session identifier
            request_id: Unique request identifier
            prompt_text: Optional prompt text (will be sanitized)
            response_text: Optional response text (will be sanitized)
            metadata: Additional metadata

        Returns:
            Telemetry event dictionary
        """
        if request_id is None:
            request_id = f"req-{int(time.time() * 1000)}-{random.randint(1000, 9999)}"

        event = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "service_name": service_name,
            "model_name": model_name,
            "latency_ms": latency_ms,
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "total_tokens": prompt_tokens + completion_tokens,
            "cost_usd": cost_usd,
            "user_id": user_id,
            "session_id": session_id,
            "request_id": request_id,
        }

        if prompt_text:
            event["prompt_text"] = prompt_text

        if response_text:
            event["response_text"] = response_text

        if metadata:
            event["metadata"] = metadata

        return event

    def send_event(self, event: Dict[str, Any]) -> None:
        """Send a telemetry event to Kafka.

        Args:
            event: Telemetry event dictionary
        """
        try:
            future = self.producer.send(self.topic, value=event)
            record_metadata = future.get(timeout=10)
            logger.info(
                f"Sent event {event['request_id']} to {record_metadata.topic} "
                f"partition {record_metadata.partition} offset {record_metadata.offset}"
            )
        except Exception as e:
            logger.error(f"Failed to send event: {e}")
            raise

    def close(self):
        """Close the producer and flush pending messages."""
        self.producer.flush()
        self.producer.close()
        logger.info("Producer closed")


def simulate_normal_traffic(producer: LLMTelemetryProducer, num_events: int = 10):
    """Simulate normal LLM traffic patterns.

    Args:
        producer: Telemetry producer instance
        num_events: Number of events to generate
    """
    logger.info(f"Simulating {num_events} normal traffic events...")

    models = ["gpt-4", "gpt-3.5-turbo", "claude-3-opus", "claude-3-sonnet"]
    services = ["chat-api", "completion-api", "assistant-api"]

    for i in range(num_events):
        # Normal latency: 500-3000ms
        latency_ms = random.uniform(500, 3000)

        # Normal token counts
        prompt_tokens = random.randint(50, 500)
        completion_tokens = random.randint(100, 800)

        # Calculate cost (example pricing)
        model = random.choice(models)
        if "gpt-4" in model:
            cost_usd = (prompt_tokens * 0.00003 + completion_tokens * 0.00006)
        else:
            cost_usd = (prompt_tokens * 0.000001 + completion_tokens * 0.000002)

        event = producer.create_telemetry_event(
            service_name=random.choice(services),
            model_name=model,
            latency_ms=latency_ms,
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            cost_usd=cost_usd,
            user_id=f"user-{random.randint(1, 100)}",
            session_id=f"session-{random.randint(1, 50)}",
            metadata={
                "region": random.choice(["us-east-1", "us-west-2", "eu-west-1"]),
                "api_version": "v1",
            }
        )

        producer.send_event(event)
        time.sleep(0.1)  # 100ms between events


def simulate_anomalous_traffic(producer: LLMTelemetryProducer, num_events: int = 5):
    """Simulate anomalous LLM traffic patterns.

    Args:
        producer: Telemetry producer instance
        num_events: Number of anomalous events to generate
    """
    logger.info(f"Simulating {num_events} anomalous traffic events...")

    anomaly_types = [
        ("high_latency", "Extremely high latency"),
        ("high_tokens", "Unusually high token count"),
        ("high_cost", "Abnormally high cost"),
        ("suspicious_pattern", "Suspicious usage pattern"),
    ]

    for i in range(num_events):
        anomaly_type, description = random.choice(anomaly_types)

        if anomaly_type == "high_latency":
            # Anomalous: 20-60 seconds
            latency_ms = random.uniform(20000, 60000)
            prompt_tokens = random.randint(100, 500)
            completion_tokens = random.randint(200, 800)
        elif anomaly_type == "high_tokens":
            # Anomalous: very high token count
            latency_ms = random.uniform(5000, 15000)
            prompt_tokens = random.randint(5000, 15000)
            completion_tokens = random.randint(8000, 20000)
        elif anomaly_type == "high_cost":
            # Anomalous: extremely high cost
            latency_ms = random.uniform(8000, 20000)
            prompt_tokens = random.randint(8000, 15000)
            completion_tokens = random.randint(10000, 25000)
        else:  # suspicious_pattern
            # Multiple rapid requests from same user
            latency_ms = random.uniform(1000, 3000)
            prompt_tokens = random.randint(50, 200)
            completion_tokens = random.randint(50, 200)

        cost_usd = (prompt_tokens * 0.00003 + completion_tokens * 0.00006)

        event = producer.create_telemetry_event(
            service_name="chat-api",
            model_name="gpt-4",
            latency_ms=latency_ms,
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            cost_usd=cost_usd,
            user_id="user-suspicious",
            session_id=f"session-anomaly-{i}",
            metadata={
                "anomaly_type": anomaly_type,
                "description": description,
                "simulated": True,
            }
        )

        producer.send_event(event)
        logger.warning(f"Sent anomalous event: {anomaly_type}")
        time.sleep(0.5)


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(description="LLM-Sentinel telemetry producer example")
    parser.add_argument(
        "--brokers",
        default="localhost:9092",
        help="Comma-separated list of Kafka brokers (default: localhost:9092)"
    )
    parser.add_argument(
        "--topic",
        default="llm.telemetry",
        help="Kafka topic name (default: llm.telemetry)"
    )
    parser.add_argument(
        "--normal-events",
        type=int,
        default=20,
        help="Number of normal events to generate (default: 20)"
    )
    parser.add_argument(
        "--anomalous-events",
        type=int,
        default=5,
        help="Number of anomalous events to generate (default: 5)"
    )
    parser.add_argument(
        "--continuous",
        action="store_true",
        help="Run continuously"
    )

    args = parser.parse_args()

    brokers = args.brokers.split(",")
    producer = LLMTelemetryProducer(brokers=brokers, topic=args.topic)

    try:
        if args.continuous:
            logger.info("Running in continuous mode (Ctrl+C to stop)...")
            while True:
                simulate_normal_traffic(producer, args.normal_events)
                simulate_anomalous_traffic(producer, args.anomalous_events)
                logger.info("Waiting 10 seconds before next batch...")
                time.sleep(10)
        else:
            simulate_normal_traffic(producer, args.normal_events)
            simulate_anomalous_traffic(producer, args.anomalous_events)
            logger.info("Finished generating events")

    except KeyboardInterrupt:
        logger.info("Interrupted by user")
    finally:
        producer.close()


if __name__ == "__main__":
    main()
